GuilhermeMota93/woocommerce-android
# Feat/products 2707 simple product details add images

## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION

**EDIT**:
This PR will contain the following:
- added the **Add image** flow for the product **"add"** path

As mentioned in [this PR](https://github.com/woocommerce/woocommerce-android/pull/2734), the idea is to reuse the product details flows as they are needed - in this case, for the **"add image"** step.

Keeping the same `isAddProduct` approach over actions, which it will let us know which path to follow on each step, I added that said "flag" on the actions for the `ProductImagesFragment`. Given this new flag, the image fragment will know to trigger either the **"edit"** upload of images, or the **"add"** one. 👍 

They are different because:
- from a previous implementation, and as it stands, the **"edit"** will upload, given an existing product's `remoteId`, the images via the `ProductsImageService` and more specifically the `MediaActionBuilder.newUploadMediaAction`;
- the **"add"** will keep also upload the images, but there won't be a `remoteProductId` to retrieve the updated data. For this, I added a couple of local "helpers" that will keep the state updated (and easy accessible) during the actual `PUBLISH` step.

Focusing on the **"add"** path/flow, I added a few things to tie everything together:
- a new view state parameter, named `addProductLocalUris` that will contain the current selected local Uri's for the add product flow;
- on the `OnProductImageUploaded`, which is the event that triggers each time a new image is uploaded and contains a remote URL, I update a local list of remote URLs in a variable called `productAddImagesRemoteUrl` so it can be easy accessible for the `PUBLISH` step;
- on the `OnProductImagesUpdateCompletedEvent`, which as the name says will trigger once the images upload is finished, I added a new field called `localUriList`. This event, on the **"add"** path, calls the `loadLocalProduct` method that updates the view states and triggers observers for UI update;
- a new type of image on the `WCProductImageGalleryView` called `VIEW_TYPE_ADD_PRODUCT_IMAGE`. This view, holds the adapter and is called to update the list items with the correct images - this new types was needed so that we can tell `Glide` to load the selected images quicker 👌 

**NOTE:**
2 caveats as it stands:
- the flow of loading an image from the `WordPress media library` is still not working as expected, I'll have to investigate it further in a tidy up PR possibly
- the click action on the images for the **"add"** path is not working, also need to investigate and fix in a follow up PR


## TESTING
- Click on the PRODUCT bottom nav bar button
    - Wait for the screen to finish loading
    - Click on the FAB button
    - Notice the product detail for a new product flow, empty with some default views
    - Click top image container
    - Notice the Image Picker screen showing
    - Select `Add Photos`
    - Notice the select input alert/dialog
 - Scenario 1:
    - Select `Choose from device`
    - Notice the camera/device roll displaying the available images
    - Select an image
 - Scenario 2:
    - Selected `Take photo`
    - Notice the camera opening to take a photo
    - Take and select photo
- Notice the image showing in the list
- Press back to navigate to the Product Detail screen 
- Notice the horizontal list with the selected images displaying

## SCREENSHOTS

<image src="https://user-images.githubusercontent.com/7058393/90697702-e426fc80-e276-11ea-82d1-4060547eab5c.gif" width="320"/> <image src="https://user-images.githubusercontent.com/7058393/90455695-a0a18680-e0ee-11ea-9277-51cf19edd9ef.gif" width="320"/> 

<image src="https://user-images.githubusercontent.com/7058393/90455820-fbd37900-e0ee-11ea-8ef2-50e9809b992a.gif" width="320"/> <image src="https://user-images.githubusercontent.com/7058393/90455792-e6f6e580-e0ee-11ea-83ea-2b883c0d0735.gif" width="320"/>
